File: javascript/032.js
/**
 * @param {string} s
 * @return {number}
 */
var longestValidParentheses = function(s) {
  let stack = [-1];

  let last = array => array[array.length - 1];

  let maxLength = 0;
  for (let i = 0; i < s.length; ++i) {
    const char = s[i];
    if (char === "(") {
      stack.push(i);
    } else {
      if (s[last(stack)] === "(") {
        stack.pop();
        const currentMaxLength = i - last(stack);
        if (currentMaxLength > maxLength) maxLength = currentMaxLength;
      } else {
        stack.push(i);
      }
    }
  }

  return maxLength;
};

File: ruby/032.rb
# @param {String} s
# @return {Integer}
def longest_valid_parentheses(s)
  stack = [-1]
  max_length = 0
  0.upto(s.length - 1).each do |i|
    if s[i] == "("
      stack.push(i)
    else
      if stack[-1] >= 0 && s[stack[-1]] == "("
        stack.pop
        current_length = i - stack[-1]
        max_length = current_length if current_length > max_length
      else
        stack.push(i)
      end
    end
  end

  max_length
end